### PR TITLE
CT: Add hash256

### PR DESCRIPTION
**Pull request by facuMH**
This PR adds an immutable `hash256` type, analogous to `Bytes` to improve performance on state operations.

For instance in `Clone`
```
facu@:~/Tosca$ benchstat oldHashClone.log newHashClone.log
goos: linux
goarch: amd64
pkg: github.com/Fantom-foundation/Tosca/go/ct/st
cpu: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
                    │ oldHashClone.log │          newHashClone.log           │
                    │      sec/op      │   sec/op     vs base                │
State_CloneState-16        9.839µ ± 5%   7.832µ ± 4%  -20.40% (p=0.000 n=25)

                    │ oldHashClone.log │           newHashClone.log           │
                    │       B/op       │     B/op      vs base                │
State_CloneState-16       41.67Ki ± 0%   32.91Ki ± 0%  -21.02% (p=0.000 n=25)

                    │ oldHashClone.log │        newHashClone.log        │
                    │    allocs/op     │ allocs/op   vs base            │
State_CloneState-16         13.00 ± 0%   13.00 ± 0%  ~ (p=1.000 n=25) ¹
¹ all samples are equal
```

And for `Eq`
```
facu@:~/Tosca$ benchstat oldHashEq.log newHashEq.log
goos: linux
goarch: amd64
pkg: github.com/Fantom-foundation/Tosca/go/ct/st
cpu: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics
                 │ oldHashEq.log │           newHashEq.log            │
                 │    sec/op     │   sec/op     vs base               │
State_EqState-16    1005.0n ± 3%   788.2n ± 6%  -21.57% (p=0.000 n=25)

                 │ oldHashEq.log │         newHashEq.log          │
                 │     B/op      │    B/op     vs base            │
State_EqState-16      0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=25) ¹
¹ all samples are equal

                 │ oldHashEq.log │         newHashEq.log          │
                 │   allocs/op   │ allocs/op   vs base            │
State_EqState-16      0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=25) ¹
¹ all samples are equal
```

### CODE REVIEW

**Comment by LuisPH3** [comment]
I had an idea about this, but I don't know if it is possible or clean.

If the conditions do not contain any constraint for the `recentBlockHashes`  you may leave it nill, fill it to zeroes when translating into the VM state.
 
The blockhash history is a read only construct from the point of view of the VM, I am not sure if we even need to compare the state after calling the VM. 

What do you guys think?

File: go/ct/common/hash256.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// Hash256 is an immutable slice of Hash256 that can be trivially cloned.
+type Hash256 struct {
+	data [256]vm.Hash
```
Comment (HerbertJordan) [inline go/ct/common/hash256.go:25]: Objects of this struct may be immutable, but it is not cheap to compare or copy.

The memory layout of this type is 256 vm.Hash values next to each other. When you compare them, all 8 KiB of data is compared, and when you assign them 8 KiB of data is copied.

To make it light-weight, the object should contain a pointer to the data. This pointer can be trivially cloned, restored, and in comparisons you can first check the pointer to the data. If it is the same, there is no need to compare 8 KiB. If it is different, you need to compare the data. But the idea is to have it unlikely that in this case the data is actually equivalent, and the comparison will abort after a few bytes.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/common/hash256.go (hunk 1 above)
Comment (facuMH) [reply go/ct/common/hash256.go]: changes applied, benchmarks rerun.

File: go/ct/common/hash256_test.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+func TestHash256_CanBeJsonEncoded(t *testing.T) {
+
+	zeroHashes := "["
+	for i := 0; i < 256; i++ {
+		zeroHashes += hash0
+		zeroHashes += ","
+	}
+	zeroHashes = zeroHashes[:len(zeroHashes)-1] + "]"
+
+	oneHashes := "[" + hash1 + ","
```
Comment (simonlechner) [inline go/ct/common/hash256_test.go:59]:
```suggestion
	oneHash := "[" + hash1 + ","
```

File: go/ct/common/hash256_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	b2 := b1
+
+	if !b1.Equal(b2) {
+		t.Errorf("assigned value is not equal: %v vs %v", b1, b2)
+	}
+}
+
+const (
+	hash0 = "[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]"
+	hash1 = "[1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]"
```
Comment (simonlechner) [inline go/ct/common/hash256_test.go:47]: Since they are only used in the function below their definition could be moved inside the function.

File: go/ct/common/hash256.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// Hash256 is an immutable array of Hash256 that can be trivially cloned.
+type Hash256 struct {
+	data *[256]vm.Hash
```
Comment (LuisPH3) [inline go/ct/common/hash256.go:25]: just a go question, this is immutable because the member is private and all getters return a copy by value, right?  no const data in go?

File: go/ct/common/hash256.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// Hash256 is an immutable array of Hash256 that can be trivially cloned.
+type Hash256 struct {
```
Comment (LuisPH3) [inline go/ct/common/hash256.go:24]: nitpick: The type contains more than one hash, name should be plural. 
Comment should say array of 256 Hashes as well.
Why not BlockHashHistory or LastBlockHashes?

File: go/ct/common/hash256.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	// Copy the slice into the Hash256 array
+	h.data = &hashes
+	return nil
+}
+
+func (b Hash256) Get(index uint64) vm.Hash {
+	if index >= 256 {
+		panic("index out of range")
```
Comment (LuisPH3) [inline go/ct/common/hash256.go:107]: Is this really a panic? Of course, the caller side cannot call this function out of range, but If we ever reuse this function somewhere else, the reviewer will not see this error, and we will hide the panic under a combination of runtime values. If instead you return an error and handle it in the adapter properly, this code will never panic.

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
good speedup.
please rename the hash256 abstraction.

File: go/ct/common/hash256.go (hunk 5 above)
Comment (facuMH) [reply go/ct/common/hash256.go:24]: I think those names convey meaning of a particular use, while this is just an immutable array. I would rather call the type `Hashes256` and fix the comment, rather than talk about blocks here.

File: go/ct/common/hash256.go (hunk 4 above)
Comment (facuMH) [reply go/ct/common/hash256.go:25]: exactly

File: go/ct/common/hash256_test.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+package common
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+)
+
+func TestHash256_EqualWhenContainingSameContent(t *testing.T) {
```
Comment (LuisPH3) [inline go/ct/common/hash256_test.go:22]: sorry, both filename and the unit tests names are out of sync now with the type name.
(I really don't like to write 256 in the type name. But I guess this is my problem ;) )

File: go/ct/st/serialization.go [hunk 8]
```diff
@@ -78,7 +78,7 @@ type stateSerializable struct {
 	CallJournal           *CallJournal
 	HasSelfDestructed     bool
 	SelfDestructedJournal []serializableSelfDestructEntry
-	RecentBlockHashes     [256]vm.Hash
+	RecentBlockHashes     Hashes256
```
Comment (LuisPH3) [inline go/ct/st/serialization.go:81]: Recent block hashes is a good name.

File: go/ct/spc/specification.go [hunk 9]
```diff
@@ -908,7 +908,9 @@ func getAllRules() []Rule {
 		effect: func(s *st.State) {
 			targetBlockNumber := s.Stack.Pop()
 			index := s.BlockContext.BlockNumber - targetBlockNumber.Uint64()
-			s.Stack.Push(NewU256FromBytes(s.RecentBlockHashes[index-1][:]...))
+			// we don't need to check for index ∈ [0..255] because InRange256FromCurrentBlock already does that
```
Comment (LuisPH3) [inline go/ct/spc/specification.go:911]: until we change the implementation and we silently break this code.
no catch all, no ignore err, plz

**Review by LuisPH3 — CHANGES_REQUESTED** [review]

File: go/ct/spc/specification.go (hunk 9 above)
Comment (LuisPH3) [reply go/ct/spc/specification.go:911]: As discussed offline, CT does not run in client and cannot produce failure there. It is kind of ok to ignore it here.

File: go/ct/common/hashes256.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+const ErrorIndexOutOfRange = ConstErr("index out of range")
+
+// Hashes256 is an immutable array of 256 vm.Hash that can be trivially cloned/compared.
+type Hashes256 struct {
+	data *[256]vm.Hash
+}
+
+func NewHash256(hashes ...vm.Hash) Hashes256 {
```
Comment (LuisPH3) [inline go/ct/common/hashes256.go:30]: name :(

File: go/ct/common/hashes256.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	for i := 0; i < 256; i++ {
+		if b.data[i] != other.data[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func Randomhash256(rnd *rand.Rand) Hashes256 {
```
Comment (LuisPH3) [inline go/ct/common/hashes256.go:51]: name

File: go/ct/common/hashes256_test.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+package common
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+)
+
+func TestHash256_EqualWhenContainingSameContent(t *testing.T) {
```
Comment (LuisPH3) [inline go/ct/common/hashes256_test.go:22]: name

File: go/ct/common/hashes256_test.go [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+	if !b1.Equal(b2) {
+		t.Errorf("instances are not equal, got %v and %v", b1, b2)
+	}
+
+	if b1.Equal(b3) {
+		t.Errorf("instances are equal, got %v and %v", b1, b3)
+	}
+}
+
+func TestHash256_AssignmentProducesEqualValue(t *testing.T) {
```
Comment (LuisPH3) [inline go/ct/common/hashes256_test.go:36]: name

File: go/ct/common/hashes256_test.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+	b1 := NewHash256(vm.Hash{1})
+	b2 := b1
+
+	if !b1.Equal(b2) {
+		t.Errorf("assigned value is not equal: %v vs %v", b1, b2)
+	}
+}
+
+func TestHash256_CanBeJsonEncoded(t *testing.T) {
```
Comment (LuisPH3) [inline go/ct/common/hashes256_test.go:45]: name

File: go/ct/common/hashes256_test.go [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+				t.Fatalf("failed to restore Hash256: %v", err)
+			}
+			if !test.hash256.Equal(restored) {
+				t.Errorf("unexpected restored value, wanted %v, got %v", test.hash256, restored)
+			}
+		})
+	}
+}
+
+func TestHash256_Get(t *testing.T) {
```
Comment (LuisPH3) [inline go/ct/common/hashes256_test.go:99]: name

File: go/ct/common/hashes256_test.go [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+	oneHash := vm.Hash{1}
+	hash256 := NewHash256(oneHash)
```
Comment (LuisPH3) [inline go/ct/common/hashes256_test.go:102]: variable name

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
Unfortunately, this not quite complete. Some types and files are renamed, but unit tests, constructor and some variables keep the old name. 
This is getting quite confusing now. I have to encourage you to find a better name.

plz check comments for mismatches or confusing uses.

File: go/ct/common/hashes256.go [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// ImmutableHashArray is an immutable array of 256 vm.Hash that can be trivially cloned/compared.
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:23]: It is not true that it can be "trivially compared".

File: go/ct/common/hashes256_test.go [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+)
+
+func TestImmutableHashArray_EqualWhenContainingSameContent(t *testing.T) {
+	b1 := NewImmutableHashArray(vm.Hash{1})
+	b2 := b1
+	b3 := NewImmutableHashArray(vm.Hash{2})
+
+	if !b1.Equal(b2) {
```
Comment (HerbertJordan) [inline go/ct/common/hashes256_test.go:27]: The test is named "EqualWhenContainngSameContent" but what is tested is only the case where the content is the same pointer. You should also make sure that two independent instances created by two distinct calls to `NewImmutableHashArray` with the same arguments are equal.

File: go/ct/common/hashes256.go
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:1]: As remarked by Luis, the file has not been aligned with the new name.

File: go/ct/common/hashes256.go [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	if b.data == nil || other.data == nil {
+		return false
+	}
+	for i := 0; i < 256; i++ {
+		if b.data[i] != other.data[i] {
+			return false
+		}
+	}
+	return true
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:46]: Shorter:
```suggestion
	return b.data == other.data || (b.data != nil && other.data != nil && *b.data == *other.data)
```

File: go/ct/common/hashes256.go [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+	return bytes
+}
+
+func (b ImmutableHashArray) FromBytes(bytes []byte) ImmutableHashArray {
+	hashes := ImmutableHashArray{}
+	for i := 0; i < 256; i++ {
+		copy(hashes.data[i][:], bytes[i*32:(i+1)*32])
+	}
+	return hashes
+}
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:72]: These two functions are nowhere used, needed, nor tested.

File: go/ct/common/hashes256.go [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+
+func (b ImmutableHashArray) ToHash() [256]vm.Hash {
+	return [256]vm.Hash(*b.data)
+}
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:76]: This function is also nowhere used nor tested.

File: go/ct/common/hashes256.go [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+func (b ImmutableHashArray) FromBytes(bytes []byte) ImmutableHashArray {
+	hashes := ImmutableHashArray{}
+	for i := 0; i < 256; i++ {
+		copy(hashes.data[i][:], bytes[i*32:(i+1)*32])
+	}
+	return hashes
+}
+
+func (b ImmutableHashArray) ToHash() [256]vm.Hash {
+	return [256]vm.Hash(*b.data)
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:75]:
```suggestion
	return *b.data
```

File: go/ct/common/hashes256.go [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+
+func (b ImmutableHashArray) String() string {
+	return fmt.Sprintf("%x", b.data)
+}
+
+func (b ImmutableHashArray) MarshalJSON() ([]byte, error) {
+	return json.Marshal(b.data)
+}
+
+// Implement the UnmarshalJSON method for ImmutableHashArray
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:86]: The comment is not formatted according to the general Go convention and does not provide any information that would not be obvious from the existence of this function.

File: go/ct/common/hashes256.go [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	// Copy the slice into the ImmutableHashArray array
+	h.data = &hashes
+	return nil
+}
+
+// Get returns the hash at the given index or panics if out of range
+func (b ImmutableHashArray) Get(index uint64) vm.Hash {
+	return vm.Hash(b.data[index])
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:107]: b.data might be `nil`, which leads to a panic;

File: go/ct/common/hashes256.go [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+	for i := 0; i < 256; i++ {
+		bytes = append(bytes, b.data[i][:]...)
+	}
+	return bytes
+}
+
+func (b ImmutableHashArray) FromBytes(bytes []byte) ImmutableHashArray {
+	hashes := ImmutableHashArray{}
+	for i := 0; i < 256; i++ {
+		copy(hashes.data[i][:], bytes[i*32:(i+1)*32])
```
Comment (HerbertJordan) [inline go/ct/common/hashes256.go:69]: This most likely crashes because `hashes.data` is `nil`.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
There are still lots of things to improve in this PR. See comments.

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/common/immutable_hash_array_test.go [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+	}{
+		"empty": {
+			NewImmutableHashArray(), NewImmutableHashArray(), true,
+		},
+		"single": {
+			NewImmutableHashArray(vm.Hash{1}), NewImmutableHashArray(vm.Hash{1}), true,
+		},
+		"diff": {
+			NewImmutableHashArray(vm.Hash{1}), NewImmutableHashArray(vm.Hash{2}), false,
+		},
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array_test.go:37]: There should be test cases with default values where the `data` field is `nil`.

File: go/ct/common/immutable_hash_array.go [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+
+// Get returns the hash at the given index or panics if out of range
+func (b ImmutableHashArray) Get(index uint64) vm.Hash {
+	if b.data == nil {
+		return vm.Hash{}
+	}
+	return vm.Hash(b.data[index])
+}
+
+func GetRandomHash(rnd *rand.Rand) vm.Hash {
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array.go:81]: nit: nobody would probably look for a function like this here. Consider moving it into its own `hash.go` file in this directory.

File: go/ct/common/immutable_hash_array_test.go [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+		oneHash += ","
+	}
+	oneHash = oneHash[:len(oneHash)-1] + "]"
+
+	tests := map[string]struct {
+		hashes  ImmutableHashArray
+		encoded string
+	}{
+		"empty": {
+			NewImmutableHashArray(vm.Hash{}), zeroHash,
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array_test.go:83]: You should also test a default value with `data == nil`.

File: go/ct/common/immutable_hash_array_test.go [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+				t.Fatalf("failed to restore ImmutableHashArray: %v", err)
+			}
+			if !test.hashes.Equal(restored) {
+				t.Errorf("unexpected restored value, wanted %v, got %v", test.hashes, restored)
+			}
+		})
+	}
+}
+
+func TestImmutableHashArray_Get(t *testing.T) {
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array_test.go:112]: Also here, the default value should be tested.

Note, the way you implemented right now, the `ImmutableHashArray` these two instances behave differently:
```
a := ImmutableHashArray{}
b := NewImmutableHashArray{}
a.Get(256) // => returns 0
b.Get(256)  // => panic
```

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/common/immutable_hash_array.go [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	return nil
+}
+
+// Get returns the hash at the given index or panics if out of range
+func (b ImmutableHashArray) Get(index uint64) vm.Hash {
+	if b.data == nil && index < 256 {
+		return vm.Hash{}
+	} else if index >= 256 {
+		panic(fmt.Sprintf("index out of range: %d", index))
```
Comment (LuisPH3) [inline go/ct/common/immutable_hash_array.go:89]: nit: this panic is redundant, go checks slices anyway.

File: go/ct/common/immutable_hash_array.go [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+	if b.data == nil {
+		return *other.data == [256]vm.Hash{}
+	}
+	if other.data == nil {
+		return *b.data == [256]vm.Hash{}
+	}
+	return b.data == other.data || (b.data != nil && other.data != nil && *b.data == *other.data)
+}
+
+func RandomImmutableHashArray(rnd *rand.Rand) ImmutableHashArray {
```
Comment (LuisPH3) [inline go/ct/common/immutable_hash_array.go:45]: nit: this is a constructor, NewRandomImmutableHashArray

File: go/ct/common/immutable_hash_array.go [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+	} else {
+		h.data = &hashes
+	}
+	return nil
+}
+
+// Get returns the hash at the given index or panics if out of range
+func (b ImmutableHashArray) Get(index uint64) vm.Hash {
+	if b.data == nil && index < 256 {
+		return vm.Hash{}
```
Comment (LuisPH3) [inline go/ct/common/immutable_hash_array.go:87]: If we panic with an out-of-bounds access, we should not silently allow calling an object that is wrongly initialized. 
for consistency, we should panic here as well.

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
Please check the error handling of the Get function.

File: go/ct/common/immutable_hash_array.go (hunk 30 above)
Comment (facuMH) [reply go/ct/common/immutable_hash_array.go:89]: discussed offline, will add documentation

File: go/ct/common/immutable_hash_array.go (hunk 32 above)
Comment (facuMH) [reply go/ct/common/immutable_hash_array.go:87]: discussed offline, will add documentationv

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/common/immutable_hash_array.go [hunk 33]
```diff
@@ -0,0 +1,94 @@
...
+	} else if index >= 256 {
+		// if data is nil, we still want to return ouf of range error
+		panic(fmt.Sprintf("index out of range: %d", index))
+	}
+	return vm.Hash(b.data[index])
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array.go:93]:
```suggestion
	return b.data[index]
```

File: go/ct/common/immutable_hash_array.go [hunk 34]
```diff
@@ -0,0 +1,94 @@
...
+		h.data = &hashes
+	}
+	return nil
+}
+
+// Get returns the hash at the given index or panics if out of range
+func (b ImmutableHashArray) Get(index uint64) vm.Hash {
+	if b.data == nil && index < 256 {
+		return vm.Hash{}
+	} else if index >= 256 {
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array.go:89]: nit: if a branch ends with a return, you can skip the `else`

File: go/ct/st/state.go [hunk 35]
```diff
@@ -457,9 +458,13 @@ func (s *State) Diff(o *State) []string {
...
 
-	for i, want := range s.RecentBlockHashes {
-		if want != o.RecentBlockHashes[i] {
-			res = append(res, fmt.Sprintf("Different block number hash at index %d: %x vs %x", i, want, o.RecentBlockHashes[i]))
+	if !s.RecentBlockHashes.Equal(o.RecentBlockHashes) {
+		for i := 0; i < 256; i++ {
+			want := s.RecentBlockHashes.Get(uint64(i))
+			got := o.RecentBlockHashes.Get(uint64(i))
+			if want != got {
+				res = append(res, fmt.Sprintf("Different block number hash at index %d: %x vs %x", i, want, got))
```
Comment (HerbertJordan) [inline go/ct/st/state.go:466]: nit:
```suggestion
				res = append(res, fmt.Sprintf("Different block hash at index %d: %x vs %x", i, want, got))
```

**Review by HerbertJordan — APPROVED** [review]
Looks good now, thanks for the improvements.

**Review by LuisPH3 — DISMISSED** [review]
alright!

File: go/ct/common/immutable_hash_array_test.go [hunk 36]
```diff
@@ -0,0 +1,144 @@
...
+		zeroHash += ","
+	}
+	zeroHash = zeroHash[:len(zeroHash)-1] + "]"
+
+	oneHash := "[" + hash1 + ","
+	for i := 0; i < 255; i++ {
+		oneHash += hash0
+		oneHash += ","
+	}
+	oneHash = oneHash[:len(oneHash)-1] + "]"
```
Comment (simonlechner) [inline go/ct/common/immutable_hash_array_test.go:80]:
```suggestion
	zeroHash := "[" + strings.Repeat(hash0+",", 255)
	zeroHash += hash0 + "]"

	oneHash := "[" + hash1 + ","
	oneHash += strings.Repeat(hash0+",", 254)
	oneHash += hash0 + "]"
```

File: go/ct/common/immutable_hash_array.go [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return err
+	}
+
+	// Check the length of the ImmutableHashArray data
+	if len(hashes) != len(h.data) {
+		return fmt.Errorf("invalid ImmutableHashArray length")
+	}
+
+	// Copy the slice into the ImmutableHashArray array
```
Comment (simonlechner) [inline go/ct/common/immutable_hash_array.go:76]:
```suggestion
	// Copy the slice into the ImmutableHashArray
```

**Review by simonlechner — DISMISSED** [review]
Looks good, just some picky suggestions.

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/common/immutable_hash_array.go [hunk 38]
```diff
@@ -0,0 +1,95 @@
...
+	if b.data == nil {
+		return *other.data == [256]vm.Hash{}
+	}
+	if other.data == nil {
+		return *b.data == [256]vm.Hash{}
+	}
+	return b.data == other.data || (b.data != nil && other.data != nil && *b.data == *other.data)
+}
+
+func NewRandomImmutableHashArray(rnd *rand.Rand) ImmutableHashArray {
```
Comment (HerbertJordan) [inline go/ct/common/immutable_hash_array.go:46]: nit: keep factory functions grouped

**Review by HerbertJordan — APPROVED** [review]

**Review by HerbertJordan — APPROVED** [review]

**Review by LuisPH3 — APPROVED** [review]
